internal/provider/resource_certificate_authority.go

```diff
@@ -0,0 +1,301 @@
+	"github.com/vmware/vcf-sdk-go/client/certificates"
+	"github.com/vmware/vcf-sdk-go/models"
+	"time"
+)
+
+func ResourceCertificateAuthority() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceCertificateAuthorityCreate,
+		ReadContext:   resourceCertificateAuthorityRead,
+		UpdateContext: resourceCertificateAuthorityUpdate,
+		DeleteContext: resourceCertificateAuthorityDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(20 * time.Minute),
+			Read:   schema.DefaultTimeout(20 * time.Minute),
+			Update: schema.DefaultTimeout(20 * time.Minute),
+			Delete: schema.DefaultTimeout(20 * time.Minute),
+		},
+		CustomizeDiff: validateRequiredAttributesForCertificateAuthority,
+		Schema: map[string]*schema.Schema{
+			"type": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Certificate authority type. \"id\" has the same value. Microsoft or OpenSSL",
+			},
+			"microsoft": {
+				Type:          schema.TypeList,
+				MaxItems:      1,
+				Optional:      true,
+				Description:   "",
+				ConflictsWith: []string{"open_ssl"},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"server_url": {
+							Type:         schema.TypeString,
+							Required:     true,
+							Description:  "Microsoft CA server URL",
+							ValidateFunc: validation.StringIsNotEmpty,
+						},
+						"template_name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							Description:  "Microsoft CA server template name",
+							ValidateFunc: validation.StringIsNotEmpty,
+						},
+						"username": {
+							Type:         schema.TypeString,
+							Required:     true,
+							Description:  "Microsoft CA server username",
+							ValidateFunc: validation.StringIsNotEmpty,
+						},
+						"secret": {
+							Type:         schema.TypeString,
+							Required:     true,
+							Sensitive:    true,
+							Description:  "Microsoft CA server password",
+							ValidateFunc: validation.StringIsNotEmpty,
+						},
+					},
+				},
+			},
+			"open_ssl": {
+				Type:          schema.TypeList,
+				MaxItems:      1,
+				Optional:      true,
+				Description:   "",
+				ConflictsWith: []string{"microsoft"},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"common_name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							Description:  "OpenSSL CA domain name",
+							ValidateFunc: validation.StringIsNotEmpty,
+						},
+						"country": {
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "ISO 3166 country code where company is legally registered",
+							ValidateFunc: validation.StringInSlice([]string{"US", "CA", "AX", "AD", "AE", "AF", "AG", "AI", "AL", "AM", "AN", "AO", "AQ", "AR", "AS", "AT", "AU",
```

Comment:
Could you please extract the country codes array to some file where constants are kept.

Reply:
Unfortunately in Go []string cannot be a constant, thus I've created GetIso3316CountryCodes() function in the constants package